### ISSUE TITLE
Suggestion to vendor botocore

### ISSUE THREAD

**Issue by rmariano** [aio-libs/aiobotocore#626]
As other users (https://github.com/aio-libs/aiobotocore/issues/624) we also faced issues with transitive dependencies (the constraints put on ``botocore`` collide with other dependencies in our projects).

If instead of depending on a narrow range of versions of ``botocore`` this one is included inside this repository (via a git sub-module, or similar approaches), it will be much simpler to handle.

Would you consider this suggestion? I volunteer to contribute.

**Comment by thehesiod** [comment]
sure, we always welcome PRs :)

**Comment by thehesiod** [comment]
but it would have to not only be botocore, but aiohttp and all the sub-dependencies no?  Not so simple, and not easy to maintain.

**Comment by rmariano** [comment]
I wonder if this can be done with a git sub-module, and in that case, we can get the updates by ``git pull`` and pointing to the new commit/tag. It would be the same as updating the version number in the ``setup.py``, only that done via git.
For the record, I also don't like vendoring dependencies, but depending on a very specific version makes it harder for clients to install alongside other packages.

**Comment by asvetlov** [comment]
I think that vendoring is a bad idea.
It overcomplicates the library maintaining.
Python doesn't support pointing the imported package version.
Packages are installed globally. What version should be used by `import botocore` statement?
Transitive dependencies make the situation is even harder. 
`botocore` depends on `python-dateutil` for example, `aiobotocore` depends on `aiohttp` which depends on `multidict`. What version of `multidict` should be used by `import multidict`?

**Comment by thehesiod** [comment]
@rmariano that's what extras are for, I suggest reading up on it, see: https://github.com/aio-libs/aiobotocore/blob/master/setup.py#L27  we're open to adding other extras that depend on botocore

**Comment by thehesiod** [comment]
I'm going back to @asvetlov 's position, against vendoring, I haven't heard a solid reason yet why it's needed.

**Comment by jcugat** [comment]
I have a similar issue with one project. The main blocker is that aiobotocore has a very strict version dependency with botocore (currently 1.10.58), which makes it very hard to use with a different version of boto3 for example.

The idea to vendor botocore is to avoid this strict version dependency, so we can use them independently. The rest of dependencies (or transitive ones) don't need to be vendored since they don't pin to specific versions of these packages.

**Comment by thehesiod** [comment]
@jcugat what's preventing you from specifying boto3 via extras?  ex: `pip install aiobotocore[boto3]`

**Comment by jcugat** [comment]
That I can only use a very specific version of boto3, in this case `1.7.58`, since boto3 has in their setup.py[0]:

```
requires = [
    'botocore>=1.10.58,<1.11.0',
    'jmespath>=0.7.1,<1.0.0',
    's3transfer>=0.1.10,<0.2.0'
]
```

And because aiobotocore pins botocore to `botocore>=1.10.58,<1.10.59` I can't use any newer version of boto3.

For reference, that boto3 version was released a month an a half ago, so it's quite outdated already.

[0] https://github.com/boto/boto3/blob/f398914adfc2d723ae15cbb2689436024874a594/setup.py#L17

**Comment by thehesiod** [comment]
so you're saying you want to use an older aiobotocore with a new boto3?  is that really useful?  It sounds like it's better to update aiobotocore

**Comment by jcugat** [comment]
No, the other way around. I want to use a newer version of boto3, but I can't.

See this example, where I try to install the latest versions of both aiobotocore and boto3:

```
$ pip install boto3==1.7.58 aiobotocore==0.9.4
Collecting boto3==1.7.58
  Using cached https://files.pythonhosted.org/packages/b8/7c/f7213ba8cf2b660a6c0cdf055290e6df71037a60fc1db3d8e4a79b4789ee/boto3-1.7.58-py2.py3-none-any.whl
Collecting aiobotocore==0.9.4
  Using cached https://files.pythonhosted.org/packages/b2/08/1e77f28301defd4fe82e498f557c99d435915c098d020a206856b3ea9ccb/aiobotocore-0.9.4-py3-none-any.whl

[skipped output]

aiobotocore 0.9.4 has requirement botocore<1.10.59,>=1.10.58, but you'll have botocore 1.10.84 which is incompatible.
Installing collected packages: jmespath, six, python-dateutil, docutils, botocore, s3transfer, boto3, wrapt, attrs, idna, idna-ssl, async-timeout, chardet, multidict, yarl, aiohttp, aiobotocore
Successfully installed aiobotocore-0.9.4 aiohttp-3.4.1 async-timeout-3.0.0 attrs-18.1.0 boto3-1.7.58 botocore-1.10.84 chardet-3.0.4 docutils-0.14 idna-2.7 idna-ssl-1.1.0 jmespath-0.9.3 multidict-4.3.1 python-dateutil-2.7.3 s3transfer-0.1.13 six-1.11.0 wrapt-1.10.11 yarl-1.2.6
```

**Comment by jcugat** [comment]
No, the other way around. I want to use a newer version of boto3, but I can't.

See this example, where I try to install the latest versions of both aiobotocore and boto3:

```
$ pip install boto3==1.8.3 aiobotocore==0.9.4
Collecting boto3==1.8.3
  Using cached https://files.pythonhosted.org/packages/5c/d4/5dc7ec1a9ed198ea88f9a10092e3577715b374e15e0c9b44589c97fbea8e/boto3-1.8.3-py2.py3-none-any.whl
Collecting aiobotocore==0.9.4
  Using cached https://files.pythonhosted.org/packages/b2/08/1e77f28301defd4fe82e498f557c99d435915c098d020a206856b3ea9ccb/aiobotocore-0.9.4-py3-none-any.whl

[skipped output]

aiobotocore 0.9.4 has requirement botocore<1.10.59,>=1.10.58, but you'll have botocore 1.11.3 which is incompatible.
Installing collected packages: six, python-dateutil, docutils, jmespath, urllib3, botocore, s3transfer, boto3, wrapt, idna, idna-ssl, chardet, attrs, async-timeout, multidict, yarl, aiohttp, aiobotocore
Successfully installed aiobotocore-0.9.4 aiohttp-3.4.1 async-timeout-3.0.0 attrs-18.1.0 boto3-1.8.3 botocore-1.11.3 chardet-3.0.4 docutils-0.14 idna-2.7 idna-ssl-1.1.0 jmespath-0.9.3 multidict-4.3.1 python-dateutil-2.7.3 s3transfer-0.1.13 six-1.11.0 urllib3-1.23 wrapt-1.10.11 yarl-1.2.6
```

**Comment by thehesiod** [comment]
ok I should clarify, an aiobotocore with an older botocore, with a boto3 with a newer botocore.  Why would you need a boto3 with newer botocore APIs, and not need then in aiobotocore as well?  Trying to understand the use case

**Comment by rmariano** [comment]
The use case would be creating a library that needs ``aiobotocore``. This will bring a ``botocore`` in only one version:

```
aiobotocore==0.9.4
botocore==1.10.58
```

Now a client imports our library, but also other dependencies, for instance ``boto3``. If it installs the latest ``boto3``, it will bring a newer version of ``botocore`` which is incompatible with the one that ``aiobotocore`` requires. The resulting constraints mean that the dependency cannot be resolved (``botocore>=1.11.4,<1.12.0',`` vs. ``1.10.58``):

https://github.com/boto/boto3/blob/7922c43946320ce0687365aeab29caf86766f49c/setup.py#L17

That means that our use of ``aiobotocore`` forces users to a particular version of ``boto3`` which they cannot change.

That's why there are suggestions of widen the range of ``botocore`` versions, or vendoring it (so it's completely independent from whatever other packages are installed).

@thehesiod maybe, like you mention, we could add a new extra for this.

**Comment by svirpridon** [comment]
Isn't locking down to a single patch an anti-pattern for semver? I'm considering dropping aiobotocore and using plain old (aio)http requests because of dependency conflicts between aiobotocore, boto3, moto, and internal packages. Nobody else is doing the patch version pinning -- it's aiobotocore doing it and breaking things.

**Comment by asvetlov** [comment]
Say again, aiobotocore is too coupled with botocore itself.
They way as the integration is done doesn't guarantee any forward and backward compatibility.
That's why aiobotocore pins the specific botocore tested and checked version.
Sorry for inconvinience

**Comment by thehesiod** [comment]
@svirpridon that's why we have those in extras
